botocore/configprovider.py

```diff
@@ -0,0 +1,416 @@
+            'api_versions',
+            config_property='api_versions',
+            default={},
+        ),
+        'credentials_file': chain_builder.build_config_chain(
+            'credentials_file',
+            env_vars='AWS_SHARED_CREDENTIALS_FILE',
+            default='~/.aws/credentials',
+        ),
+        'metadata_service_timeout': chain_builder.build_config_chain(
+            'metadata_service_timeout',
+            env_vars='AWS_METADATA_SERVICE_TIMEOUT',
+            config_property='metadata_service_timeout',
+            default=1,
+            conversion_func=int,
+        ),
+        'metadata_service_num_attempts': chain_builder.build_config_chain(
+            'metadata_service_num_attempts',
+            env_vars='AWS_METADATA_SERVICE_NUM_ATTEMPTS',
+            config_property='metadata_service_num_attempts',
+            default=1,
+            conversion_func=int,
+        ),
+        'parameter_validation': chain_builder.build_config_chain(
+            'parameter_validation',
+            config_property='parameter_validation',
+            default=True,
+        ),
+    }
+
+
+class ConfigChainFactory(object):
+    """Factory class to create our most common configuration chain case.
+
+    This is a convenience class to construct configuration chains that follow
+    our most common pattern. This is to prevent ordering them incorrectly,
+    and to make the config chain construction more readable.
+    """
+    def __init__(self, session, environ=None):
+        """Initialize a ConfigChainFactory.
+
+        :type session: :class:`botocore.session.Session`
+        :param session: This is the session that should be used to look up
+            values from the config file.
+
+        :type environ: dict
+        :param environ: A mapping to use for environment variables. If this
+            is not provided it will default to use os.environ.
+        """
+        self._session = session
+        if environ is None:
+            environ = os.environ
+        self._environ = environ
+
+    def build_config_chain(self, logical_name, instance=True, env_vars=None,
+                           config_property=None, default=None,
+                           conversion_func=None):
+        """Build a config chain following the standard botocore pattern.
+
+        In botocore most of our config chains follow the the precendence:
+        session_instance_variables, environment, config_file, default_value.
+
+        This is a convenience function for creating a chain that follow
+        that precendence.
+
+        :type logical_name: str
+        :param logical_name: The logical name of the config value that this
+            chain is responsible for loading.
+
+        :type instance: bool
+        :param instance: This indicates whether or not session instance
+            variable lookup should be included in the config chain. By default
+            it is True.
+
+        :type env_vars: str or list of str or None
```

Comment:
Just a little more feedback on the names after looking at the tests... Maybe call this `env_varnames` or something else to indicate it is a name? Typically, `env_vars` or `env` is associated as a map and it threw me off when I was reading through the tests. The same thing goes for `config_property`, but not as much. Maybe try to find a parameter name that better indicates it expects the `name`?